### PR TITLE
Ajax call for modal image upload

### PR DESCRIPTION
**Pull request by benlowkh**

### CODE REVIEW

File: app/lib/regex.py [hunk 1]
```diff
@@ -9,5 +9,6 @@
 
 SLUG_REGEX = r'[0-9a-zA-Z-]+'
 FILENAME_REGEX = r'[\w\-@\|\(\)]+'
+EXTENSION_REGEX = "|".join(ALLOWED_UPLOAD_EXTENSIONS)
 FULL_FILENAME_REGEX = "{fname}({ext})".format(fname=FILENAME_REGEX,
-        ext="|".join(ALLOWED_UPLOAD_EXTENSIONS))
+											  ext=EXTENSION_REGEX)
```
Comment (danrschlosser) [inline app/lib/regex.py]: please config your editor to uses spaces, not tabs!  then convert :)

**Comment by benlowkh** [comment]
@bluemelodia @EvanNGoldstein

@danrschlosser : Adam's error (https://www.dropbox.com/s/e9n9gy2evio2ey5/Screenshot%202015-03-29%2013.15.49.png?dl=0) might be a nginx error

File: app/forms/UploadImageForm.py [hunk 2]
```diff
@@ -27,7 +28,9 @@ class UploadImageForm(Form):
     image = FileField('Image file')
     uploaded_from = StringField('Uploaded from')
     filename = StringField('Filename', [
-        Regexp(FILENAME_REGEX),
+        Regexp(FILENAME_REGEX, message='Invalid filename.'),
```
Comment (danrschlosser) [inline app/forms/UploadImageForm.py]: Can we be more helpful here?  Like "Your filename should only contain ..."

File: app/forms/UploadImageForm.py [hunk 3]
```diff
@@ -27,7 +28,9 @@ class UploadImageForm(Form):
...
         Required('Please submit a filename'),
         UniqueImage()])
-    extension = StringField('Extension')
+    extension = StringField('Extension', [
+        Regexp(EXTENSION_REGEX, message='Only images are allowed.'),
```
Comment (danrschlosser) [inline app/forms/UploadImageForm.py]: Let's create something like "Only .png, .jpg, ... files are allowed."

File: app/routes/admin/media.py [hunk 4]
```diff
@@ -112,3 +109,8 @@ def delete(filename):
         flash('Invalid filename')
         pass
     return redirect(url_for('.index'))
+
+@media.route('/media/view', methods=['GET'])
+def view():
```
Comment (danrschlosser) [inline app/routes/admin/media.py]: Docstring this please!  See other similar routes

File: app/static/img/uploaded/README.md [hunk 5]
```diff
@@ -1,7 +0,0 @@
-# `/app/static/img/uploaded`
```
Comment (danrschlosser) [inline app/static/img/uploaded/README.md]: Should these changes be here?

File: app/static/js/admin/events/create.js [hunk 6]
```diff
@@ -220,6 +220,16 @@ $(function() {
         $('.display-image').addClass('hidden');
     });
 
+    /*When image modal opens, request images*/
+    $(document).on('click', 'a[href="#show-modal"][data-modal="image"]', function(e) {
+        e.preventDefault();
+        $('#images-ajax-loadpoint').load("/admin/media/view", function(response, status){
+            if (status == "error"){
+                alert("Sorry, there was an error loading the images.");
```
Comment (danrschlosser) [inline app/static/js/admin/events/create.js]: Can we display the specific error if one exists?  In general `alert()` is not a good option most of the time. I'd rather see an error message on the screen, and even better would be if the error message could be the same one that is sent from Flask

File: app/static/js/admin/media/upload.js [hunk 7]
```diff
@@ -23,8 +23,47 @@ $(function() {
 
     /* let the `btn-upload` div act as the submit button */
     $(".btn-upload").click(function () {
-        $('form.upload-form').submit();
-        return false;
+        var formData = new FormData($('.upload-form')[0]);
+        $.ajax({
+            url: '/admin/media/upload',  //server script to process data
```
Comment (danrschlosser) [inline app/static/js/admin/media/upload.js]: let's remove that comment

File: app/static/js/admin/media/upload.js [hunk 8]
```diff
@@ -23,8 +23,47 @@ $(function() {
...
+            type: 'POST',
+            xhr: function() {  // custom xhr
```
Comment (danrschlosser) [inline app/static/js/admin/media/upload.js]: Why do we do this?

File: app/static/js/admin/media/upload.js [hunk 9]
```diff
@@ -23,8 +23,47 @@ $(function() {
...
+                myXhr = $.ajaxSettings.xhr();
+                return myXhr;
+            },
+            //Ajax events
+            success: function(data, textStatus, jqXHR) {
+                var response = jQuery.parseJSON(jqXHR.responseText);
+                if (response.extension == null && response.filename == null){
```
Comment (danrschlosser) [inline app/static/js/admin/media/upload.js]: Can you explain what's happening here?  If we're in the success function, when would the extension / filename be null?

File: app/static/js/admin/media/upload.js [hunk 10]
```diff
@@ -23,8 +23,47 @@ $(function() {
...
+                    $('#images-ajax-loadpoint').load("/admin/media/view", function(response, status){
+                        if (status == "error"){
+                            $('error-message').text("Sorry, there was an error loading the images.");
+                        }
+                    });
+                    $('.error-message').hide();
+                }
+                else {
+                    if (response.extension == null)
```
Comment (danrschlosser) [inline app/static/js/admin/media/upload.js]: Why do we show populate the error message with the text and/or extension?  Is there a more meaningful error to show if we get here?

File: app/static/js/admin/media/upload.js [hunk 11]
```diff
@@ -23,8 +23,47 @@ $(function() {
...
+                }
+            },
+            error: function(jqXHR, textStatus, error) {
+                $('.error-mesage').text("An unknown error occurred.");
+                $('.error-message').addClass('active');
+            },
+            // Form data
+            data: formData,
+            //Options to tell JQuery not to process data or worry about content-type
+            cache: false,
```
Comment (danrschlosser) [inline app/static/js/admin/media/upload.js]: Might it be nice to cache this request?

File: app/static/scss/admin/media/upload.scss [hunk 12]
```diff
@@ -48,4 +48,16 @@
     .btn-choose {
         margin-right: 1rem;
     }
+    .error-message{
+        display: none;
+        background-color: $red;
+        color: $white;
+        height: 25px;
```
Comment (danrschlosser) [inline app/static/scss/admin/media/upload.scss]: Would you convert these `px` measurements to `rem`?  I think a height of `1.5rem` and padding of `0.25rem` would come pretty close to what you have now.

File: app/static/scss/lib/jquery.timepicker.scss [hunk 13]
```diff
@@ -1,4 +1,4 @@
-@import '../colors';
+@import '../admin/colors';
```
Comment (danrschlosser) [inline app/static/scss/lib/jquery.timepicker.scss]: Yeah I don't know why this has been broken for so long. We should def merge this in.

File: app/templates/admin/events/create.html [hunk 14]
```diff
@@ -100,16 +100,8 @@
...
-{% for image in images %}
-<li class="image" data-filename="{{ image.filename }}">
-    <a data-filename="{{ image.filename }}" data-url="{{ image.url() }}" href="#set-image">
-        <i style="background-image:url({{ image.url() }});"></i>
-        <div class="select"><i class="fa fa-plus fa-3x"></i></div>
-    </a>
-</li>
-{% endfor %}
-</ul>
+<div id='images-ajax-loadpoint'>        <!-- edited code -->
```
Comment (danrschlosser) [inline app/templates/admin/events/create.html]: Let's remove that comment :)

File: app/templates/macros.html [hunk 15]
```diff
@@ -76,9 +76,12 @@
     <div class="btn btn-choose">Choose File</div>
     <div class="filename-wrapper">
         {{ render_field(form.filename, label=False, placeholder="No File Chosen", required="")}}
+        <div id='ajax-loading-images'>
```
Comment (danrschlosser) [inline app/templates/macros.html]: Why do we have this as well as the `<div id='images-ajax-loadpoint'> ` in `app/templates/admin/events/create.html`?

File: config/flask_config.py [hunk 16]
```diff
@@ -75,12 +75,6 @@
...
 
 # The file extensions that may be uploaded
-ALLOWED_UPLOAD_EXTENSIONS = set(['.txt',
-    '.pdf',
-    '.png',
-    '.jpg',
-    '.jpeg',
-    '.gif'
-])
+ALLOWED_UPLOAD_EXTENSIONS = set(['.png', '.jpg', '.jpeg', '.gif'])
```
Comment (danrschlosser) [inline config/flask_config.py]: :+1:

File: eventum notes.txt [hunk 17]
```diff
@@ -0,0 +1,5 @@
+how to submit a form in ajax without a library in jquery
```
Comment (danrschlosser) [inline eventum notes.txt]: hehe let's remove this file

**Comment by danrschlosser** [comment]
Would you also include a screenshot of the error box after you take a look at some of my comments?

**Comment by natebrennand** [comment]
pping @benlowkh

**Comment by benlowkh** [comment]
still working on it! for some reason the code broke after merging travis
into my local branch, so i'm figuring out what's going on

**Comment by danrschlosser** [comment]
Hey @benlowkh I think you forgot to push the `README.md` file, would you do that and re-push?  We're :+1: once you do

**Comment by benlowkh** [comment]
Apparently the README.md file was on the gitignore list, so I force pushed it

**Pull request merged**